### PR TITLE
feat: add charm-level tracing to MAAS region and agent

### PR DESCRIPTION
**Pull request by alema-r**
This PR adds charm-level tracing to both the charms.

The tempo_k8s' libs are necessary for adding the feature
To test the integration I've added a `scenario` test.

### CODE REVIEW

**Review by wyattrees — COMMENTED** [review]
Tentative LGTM, as I haven't had time yet to test this out

File: maas-agent/src/charm.py [hunk 1]
```diff
@@ -36,6 +38,7 @@
 MAAS_SNAP_CHANNEL = "3.5/stable"
 
 
+@trace_charm(tracing_endpoint="charm_tracing_endpoint")
```
Comment (skatsaounis) [inline maas-agent/src/charm.py:41]: What about `extra_types` on the classes we setup for our integrations? `COSAgentProvider`, `MaasRegionProvider`. Did you try to add them and observe any differences in the captured traces?

File: maas-region/src/charm.py [hunk 2]
```diff
@@ -53,6 +55,7 @@
 ]
 
 
+@trace_charm(tracing_endpoint="charm_tracing_endpoint")
```
Comment (skatsaounis) [inline maas-region/src/charm.py:58]: What about `extra_types` on the classes we setup for our integrations? `COSAgentProvider`, `MaasRegionProvider`, `DatabaseRequires`, . Did you try to add them and observe any differences in the captured traces?

Comment (alema-r) [reply maas-region/src/charm.py]: By adding those classes to the `extra_types` I can see lots of more traces. I've added all of them together with `MaasHelper` that can be useful for inspecting the stacktrace when some error happens (see attached image).
Screenshot from 2024-10-07 11-04-54

File: maas-agent/src/charm.py [hunk 3]
```diff
@@ -36,6 +38,13 @@
 MAAS_SNAP_CHANNEL = "3.5/stable"
 
 
+@trace_charm(
+    tracing_endpoint="charm_tracing_endpoint",
+    extra_types=[
+        cos_agent.COSAgentProvider,
+        maas.MaasRegionProvider,
```
Comment (skatsaounis) [inline maas-agent/src/charm.py:45]: Unless I am missing something here, I think that this should be corrected:

```suggestion
        maas.MaasRegionRequirer,
```

Comment (skatsaounis) [reply maas-agent/src/charm.py:45]: Plus what about `MaasHelper` here?

**Review by skatsaounis — CHANGES_REQUESTED** [review]

File: maas-agent/src/charm.py (hunk 3 above)
Comment (alema-r) [reply maas-agent/src/charm.py]: Oh, you're definitely right. Corrected those!

**Pull request merged**